drivers/flash/flash_cadence_nand_ll.c

```diff
@@ -0,0 +1,1516 @@
+		LOG_ERR("Memory allocation error occurred %s", __func__);
+		return -ENOSR;
+	}
+	ret = cdns_cdma_desc_transfer_finish(params, cdma_desc, page_count,
+					     CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT, CNF_CMD_RD,
+					     start_page_number, buffer);
+
+	if (ret != 0) {
+		k_free(cdma_desc);
+		return ret;
+	}
+
+	k_free(cdma_desc);
+
+	return 0;
+}
+
+/**
+ * Perform a CDMA erase operation for the Cadence NAND controller.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param start_block_number The starting block number for the erase operation.
+ * @param block_count The number of blocks to be erased.
+ * @retval  0 on success or negative error value on failure.
+ */
+static int cdns_nand_cdma_erase(struct cadence_nand_params *params, uint32_t start_block_number,
+				uint32_t block_count)
+{
+	struct cdns_cdma_command_descriptor *cdma_desc;
+	int ret;
+
+	cdma_desc = cdma_allocate_desc_memory(block_count, CONFIG_FLASH_CDNS_CDMA_BLOCK_COUNT);
+
+	if (cdma_desc == NULL) {
+		LOG_ERR("Memory allocation error occurred %s", __func__);
+		return -ENOSR;
+	}
+	ret = cdns_cdma_desc_transfer_finish(params, cdma_desc, block_count,
+					     CONFIG_FLASH_CDNS_CDMA_BLOCK_COUNT, CNF_CMD_ERASE,
+					     start_block_number, NULL);
+
+	if (ret != 0) {
+		k_free(cdma_desc);
+		return ret;
+	}
+
+	k_free(cdma_desc);
+
+	return 0;
+}
+#endif
+
+#if CONFIG_CDNS_NAND_PIO_MODE
+
+/**
+ * Perform an erase operation on the Cadence NAND controller using PIO.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param thread The thread identifier for the PIO operation.
+ * @param bank The bank identifier for the erase operation.
+ * @param start_block The starting block number for the erase operation.
+ * @param ctype The command type for the erase operation.
+ * @param block_count The number of blocks to be erased.
+ * @retval   0 on success or negative error value on failure.
+ */
+static int cdns_nand_pio_erase(struct cadence_nand_params *params, uint8_t thread, uint8_t bank,
+			       uint32_t start_block, uint16_t ctype, uint32_t block_count)
+{
+	uint32_t status;
+	uintptr_t base_address;
+	uint32_t row_address = 0;
+	uint32_t index = 0;
+	int ret;
+
+	base_address = params->nand_base;
+	for (index = 0; index < block_count; index++) {
+
+		ret = cdns_wait_for_thread(base_address, thread);
+
+		if (ret != 0) {
+			return ret;
+		}
+		row_address_set(params, &row_address, (start_block * params->npages_per_block));
+		sys_write32(row_address, (base_address + CDNS_CMD_REG1));
+		start_block++;
+		sys_write32((NF_CMD4_BANK_SET(bank)), (base_address + CDNS_CMD_REG4));
+		status = CMD_0_THREAD_POS_SET(thread);
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+		status |= BIT(PIO_CF_INT);
+#endif
+		status |= CMD_0_C_MODE_SET(CT_PIO_MODE);
+		status |= PIO_CMD0_CT_SET(ctype);
+		sys_write32(status, (base_address + CDNS_CMD_REG0));
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+		k_sem_take(&params->interrupt_sem_t, K_FOREVER);
+#endif
```

Comment:
You could just have some `INT_SEM_TAKE` macro defined, that would depend on the `CONFIG_CDNS_NAND_INTERRUPT_SUPPORT`... less ifdefs cutting into code.

Reply:
macro added and replaced with new macro.